### PR TITLE
Add ignore_and_fire neuron model

### PR DESCRIPTION
**Pull request by mlober**
This PR adds the ignore_and_fire neuron model of @tomtetzlaff. This neuron model was developed for benchmarking purposes with the goal of retaining constant activity levels while up- and down-scaling of the network. It sends out spikes in fixed time intervals, independent of the given input. It is handled like a regular neuron model by nest.

I suggest @tomtetzlaff and @jessica-mitchell as reviewers.

### CODE REVIEW

File: models/ignore_and_fire.h [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
++++++++++++
+
+The ``ignore_and_fire`` neuron is a neuron model generating spikes with a fixed predefined inter-spike interval,
+irrespective of its inputs ("ignore"). In this simplest version of the ignore_and_fire neuron, the inputs from other
+neurons or devices are not processed at all (*). The ``ignore_and_fire`` neuron is primarily used for neuronal-network
+model validation and benchmarking purposes, in particular to evaluate the correctness and performance of connectivity
+generation and inter-neuron communication. It permits an easy scaling of the network size and/or connectivity without
+affecting the output spike statistics. The amount of network traffic is predefined by the user, and therefore fully
+controllable and predictable, irrespective of the network size and structure.
+
```
Comment (tomtetzlaff) [inline models/ignore_and_fire.h:54]: Here, I would suggest to add the remaining explanatory text from the ACA testcases:

"To generate asynchronous activity for a population of ``ignore_and_fire`` neurons, the firing ``phase``s can be randomly initialized. Note that the firing ``phase`` is a real number, defined as the time to the next spike relative to the firing period."

File: models/ignore_and_fire.h [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+dynamics to make it more similar and comparable to other non-ignorant neuron models. In such extended ignore_and_fire
+models, the spike emission process would still be decoupled from the intrinsic neuron dynamics.
+
+Parameters
+++++++++++
+
+The following parameters can be set in the status dictionary.
+
+================= ======= ======================================================
+ phase                     Phase
```
Comment (tomtetzlaff) [inline models/ignore_and_fire.h:65]: Add:
"Phase (relative time to next spike; 0<phase<=1)"

File: models/ignore_and_fire.h [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+ rate               1/s    Firing rate
+================= ======= ======================================================
+
+References
```
Comment (tomtetzlaff) [inline models/ignore_and_fire.h:69]: These reference have nothing to do with this model and can be removed.

File: models/ignore_and_fire.h [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+SpikeEvent
+
+Receives
+++++++++
+
+SpikeEvent, CurrentEvent, DataLoggingRequest
+
+EndUserDocs */
+
+/**
```
Comment (tomtetzlaff) [inline models/ignore_and_fire.h:93]: The following comment can be removed. There exists no propagator here and no matrix multiplications are needed.

**Review by tomtetzlaff — CHANGES_REQUESTED** [review]
Thanks a lot Melissa. Looks good to me. Just some minor things in the documentation.

**Comment by clinssen** [comment]
Hi, thank you for the contribution! Is there any unit test that you might be able to add for the model?

**Comment by tomtetzlaff** [comment]
Hi Charl, maybe one could create one from this:

```
import nest
import numpy

def ignore_and_fire_test():
     
    rate = 5.            ## firing rate (spikes/s)
    phase = 0.4          ## firing phase
    dt = 2**-3           ## simulation time resolution (ms)
    T = 1000.            ## simulation time (ms)

    nest.ResetKernel()
    nest.SetKernelStatus({
        'print_time': True,
        'resolution': dt
    })

    neuron = nest.Create("ignore_and_fire", 1)
    neuron.set({'rate': rate, 'phase': phase})

    spike_recorder = nest.Create('spike_recorder')

    nest.Connect(neuron,spike_recorder)

    nest.Simulate(T)

    ## theoretical spike times
    period = 1./rate * 1e3
    first_spike_time = phase * period + dt
    spike_times_target = numpy.arange(first_spike_time, T, period)

    ## spike times of NEST model
    spike_times_nest = nest.GetStatus(spike_recorder,'events')[0]['times']

    print()
    print('spike times (NEST):', spike_times_nest)
    print('spike times (target):', spike_times_target)

    test_passed = numpy.alltrue(spike_times_nest == spike_times_target)
    print('test passed: %r' % test_passed)
    
    return test_passed

ignore_and_fire_test()
```

File: models/ignore_and_fire.h [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+
+Short description
++++++++++++++++++
+
+Neuron generating spikes at fixed intervals irrespective of inputs.
+
+Description
++++++++++++
+
+The ``ignore_and_fire`` neuron is a neuron model generating spikes with a fixed predefined inter-spike interval,
```
Comment (tomtetzlaff) [inline models/ignore_and_fire.h:47]: After discussing with Jessica, I'd suggest the following revision (incl. paragraphs):

"
The ``ignore_and_fire`` neuron is a neuron model generating spikes at a predefined ``firing_rate`` with a constant inter-spike interval ("fire"), irrespective of its inputs ("ignore"). In this simplest version of the ``ignore_and_fire`` neuron, the inputs from other neurons or devices are not processed at all (*). 

The model's state variable, the ``phase``, describes the time to the next spike relative to the firing period (the inverse of the ``rate``). In each update step, the ``phase`` is decremented by a fixed amount. If it hits zero, a spike is emitted and the phase is reset to +1.

To create asynchronous activity for a population of ``ignore_and_fire`` neurons, the firing ``phase``s can be randomly initialized.

The ``ignore_and_fire`` neuron is primarily used for neuronal-network model verification and validation purposes ("benchmarking"), in particular, to evaluate the correctness and performance of connectivity generation and inter-neuron communication. It permits an easy scaling of the network size and/or connectivity without affecting the output spike statistics. The amount of network traffic is predefined by the user, and therefore fully controllable and predictable, irrespective of the network size and structure. 

(*) The model can easily be extended and equipped with any arbitrary input processing (such as calculating input currents with alpha-function shaped PSC kernels or updating the gating variables in the Hodgkin-Huxley model) or (after-) spike generation dynamics to make it more similar and comparable to other non-ignorant neuron models. In such extended ignore_and_fire models, the spike emission process would still be decoupled from the intrinsic neuron dynamics.
"

File: models/ignore_and_fire.h [hunk 6]
```diff
@@ -0,0 +1,309 @@
...
+currents with alpha-function shaped PSC kernels or updating the gating variables in the Hodgkin-Huxley model) or
+(after-) spike generation dynamics to make it more similar and comparable to other non-ignorant neuron models. In such
+extended ignore_and_fire models, the spike emission process would still be decoupled from the intrinsic neuron dynamics.
+
+Parameters
+++++++++++
+
+The following parameters can be set in the status dictionary.
+
+================= ======= ======================================================
```
Comment (tomtetzlaff) [inline models/ignore_and_fire.h:74]: =============== ======== =============================== ========================================================================
**Parameter**   **Unit** **Math equivalent**             **Description**
=============== ======== =============================== ========================================================================
 ``phase``        -                                      Phase (relative time to next spike; 0<``phase``<=1)
 ``rate``        1/s                                     Firing rate
=============== ======== =============================== ========================================================================

File: models/ignore_and_fire.h [hunk 7]
```diff
@@ -0,0 +1,309 @@
...
++++++++++++++++++
+
+Neuron generating spikes at fixed intervals irrespective of inputs.
+
+Description
++++++++++++
+
+The ``ignore_and_fire`` neuron is a neuron model generating spikes at a predefined ``rate`` with a constant inter-spike
+interval ("fire"), irrespective of its inputs ("ignore"). In this simplest version of the ``ignore_and_fire`` neuron,
+the inputs from other neurons or devices are not processed at all (*).
```
Comment (jessica-mitchell) [inline models/ignore_and_fire.h:49]:
```suggestion
the inputs from other neurons or devices are not processed at all.
```

File: models/ignore_and_fire.h [hunk 8]
```diff
@@ -0,0 +1,309 @@
...
+
+The model's state variable, the ``phase``, describes the time to the next spike relative to the firing period (the
+inverse of the ``rate``). In each update step, the ``phase`` is decremented by a fixed amount. If it hits zero, a spike
```
Comment (jessica-mitchell) [inline models/ignore_and_fire.h:52]:
```suggestion
inverse of the ``rate``). In each update step, the ``phase`` is decreased by a fixed amount. If it hits zero, a spike
```

File: models/ignore_and_fire.h [hunk 9]
```diff
@@ -0,0 +1,309 @@
...
+To create asynchronous activity for a population of ``ignore_and_fire`` neurons, the firing phases can be randomly
+initialized.
+
+The ``ignore_and_fire`` neuron is primarily used for neuronal-network model verification and validation purposes
+("benchmarking"), in particular, to evaluate the correctness and performance of connectivity generation and inter-neuron
+communication. It permits an easy scaling of the network size and/or connectivity without affecting the output spike
+statistics. The amount of network traffic is predefined by the user, and therefore fully controllable and predictable,
+irrespective of the network size and structure.
+
+(*) The model can easily be extended and equipped with any arbitrary input processing (such as calculating input
```
Comment (jessica-mitchell) [inline models/ignore_and_fire.h:64]:
```suggestion
.. note::

   The model can easily be extended and equipped with any arbitrary input processing (such as calculating input
```

File: models/ignore_and_fire.h [hunk 10]
```diff
@@ -0,0 +1,309 @@
...
+currents with alpha-function shaped PSC kernels or updating the gating variables in the Hodgkin-Huxley model) or
```
Comment (jessica-mitchell) [inline models/ignore_and_fire.h:65]:
```suggestion
   currents with alpha-function shaped PSC kernels or updating the gating variables in the Hodgkin-Huxley model) or
```

File: models/ignore_and_fire.h [hunk 11]
```diff
@@ -0,0 +1,309 @@
...
+(after-) spike generation dynamics to make it more similar and comparable to other non-ignorant neuron models. In such
```
Comment (jessica-mitchell) [inline models/ignore_and_fire.h:66]:
```suggestion
   (after-) spike generation dynamics to make it more similar and comparable to other non-ignorant neuron models. In such
```

File: models/ignore_and_fire.h [hunk 12]
```diff
@@ -0,0 +1,309 @@
...
+extended ignore_and_fire models, the spike emission process would still be decoupled from the intrinsic neuron dynamics.
```
Comment (jessica-mitchell) [inline models/ignore_and_fire.h:67]:
```suggestion
   extended ignore_and_fire models, the spike emission process would still be decoupled from the intrinsic neuron dynamics.
```

File: models/ignore_and_fire.h [hunk 13]
```diff
@@ -0,0 +1,309 @@
...
+Sends
++++++
+
+SpikeEvent
+
+Receives
+++++++++
+
+SpikeEvent, CurrentEvent, DataLoggingRequest
+
```
Comment (jessica-mitchell) [inline models/ignore_and_fire.h:88]:
```suggestion

Examples using this model
+++++++++++++++++++++++++

.. listexamples:: ignore_and_fire

```

I know there are no examples currently for this model, but having this directive here will make sure if we do add examples with ignore_and_fire, they will automatically be listed here.

File: models/ignore_and_fire.h (hunk 10 above)
Comment (jessica-mitchell) [reply models/ignore_and_fire.h:65]: Format the text in asterisk as a reStructured Text note

File: models/ignore_and_fire.h [hunk 14]
```diff
@@ -0,0 +1,309 @@
...
+namespace nest
+{
+
+/* BeginUserDocs: neuron, integrate-and-fire, current-based
+
+Short description
++++++++++++++++++
+
+Neuron generating spikes at fixed intervals irrespective of inputs.
```
Comment (jessica-mitchell) [inline models/ignore_and_fire.h:42]:
```suggestion
Ignore-and-fire neuron model for generating spikes at fixed intervals irrespective of inputs
```

**Review by jessica-mitchell — CHANGES_REQUESTED** [review]
thanks @mlober and @tomtetzlaff for your contribution! 
I have a few changes requested: also the table is not rendering on Read the Docs and needs to be fixed up. See output: https://nest-simulator--2987.org.readthedocs.build/en/2987/models/ignore_and_fire.html

**Review by tomtetzlaff — COMMENTED** [review]
Thanks Jessica. This looks all very good to me. How can we fix the table? I copied the format from some other neuron model, iaf_psc_alpha, I think.

File: models/ignore_and_fire.h [hunk 15]
```diff
@@ -0,0 +1,317 @@
...
+dynamics.
+
+Parameters
+++++++++++
+
+The following parameters can be set in the status dictionary.
+
+================= ======= ======================================================
+ ``phase``                     Phase (relative time to next spike; 0<phase<=1)
+ ``rate``               1/s    Firing rate
```
Comment (jessica-mitchell) [inline models/ignore_and_fire.h:79]:
```suggestion
 ``rate``            1/s       Firing rate
```

**Comment by jessica-mitchell** [comment]
> Thanks Jessica. This looks all very good to me. How can we fix the table? I copied the format from some other neuron model, iaf_psc_alpha, I think.

I fixed the table, it was only a misalignment of the second item in the second  row

File: models/ignore_and_fire.h (hunk 6 above)
Comment (jessica-mitchell) [reply models/ignore_and_fire.h:77]: I think the "math equivalent" does not need to be included here, unless an equation is going to be added

Comment (jessica-mitchell) [reply models/ignore_and_fire.h:77]:
```
================= ========= ======================================================
**Parameter**     **Unit**  **Description**
================= ========= ======================================================
 ``phase``                     Phase (relative time to next spike; 0<phase<=1)
 ``rate``          1/s         Firing rate
================= ========= ======================================================
```
@mlober can you update the model to include the headers like this?

**Review by jessica-mitchell — APPROVED** [review]
thanks @mlober @tomtetzlaff lgtm!

**Comment by jessica-mitchell** [comment]
ping @clinssen :)

**Review by clinssen — APPROVED** [review]
Great, thank you!

**Comment by heplesser** [comment]
@tomtetzlaff Could you re-review so we can proceed to merging?

**Pull request merged**